File: 0841-keys-and-rooms/0841-keys-and-rooms.cpp
class Solution {
public:

    void dfs(vector<vector<int>> &rooms, int s, vector<bool> &record){
        record[s] = true;
        for(int i = 0; i < rooms[s].size(); i++){
            if(!record[rooms[s][i]]){
                dfs(rooms, rooms[s][i], record);
            }
        }
    }

    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int i, n=rooms.size();
        vector<bool> record(n, false);
        dfs(rooms, 0, record);
        for(i = 0; i < n; i++){
            if(!record[i]){
                return false;
            }
        }
        return true;
    }
};

File: 0841-keys-and-rooms/0841-keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        
        def dfs(v):

            for i in rooms[v]:
                if not record[i]:
                    record[i] =1 
                    dfs(i)

        record= [0]*len(rooms)
        record[0] =1 
        dfs(0)
        for r in record:
            if not r:
                return False 
        return True 